support-frontend/app/controllers/RedemptionController.scala

```diff
@@ -220,19 +237,40 @@ class TestUserFromRequest(identityService: IdentityService, testUsers: TestUserS
 
 }
 
-class GetCorporateCustomer(dynamoLookup: DynamoTableAsyncForUser) {
+class CodeValidator(zuoraLookupService: ZuoraGiftLookupService, dynamoLookup: DynamoTableAsyncForUser) {
+
+  def validate(inputCode: String, isTestUser: Boolean)(implicit ec: ExecutionContext): EitherT[Future, String, ReaderType] =
+    EitherT(getValidationResult(inputCode, isTestUser).map {
+      case ValidCorporateCode(_) => Right(Corporate)
+      case ValidGiftCode(_) => Right(Gift)
+      case invalidCode => Left(validationResultToErrorMessage(invalidCode))
+    })
 
-  def apply(redemptionCode: String, isTestUser: Boolean)(implicit ec: ExecutionContext): EitherT[Future, String, Unit] = {
-    val getCodeStatus = GetCodeStatus.withDynamoLookup(dynamoLookup(isTestUser))
+  private def getValidationResult(inputCode: String, isTestUser: Boolean)(implicit ec: ExecutionContext): Future[CodeValidationResult] = {
+    val corporateValidator = CorporateCodeValidator.withDynamoLookup(dynamoLookup(isTestUser))
+    val giftValidator = new GiftCodeValidator(zuoraLookupService)
 
-    for {
-      codeToCheck <- EitherT.fromEither[Future](RedemptionCode(redemptionCode)).leftMap(_ => "Please check the code and try again")
-      _ <- EitherT(getCodeStatus(codeToCheck)).leftMap {
-        case GetCodeStatus.NoSuchCode => "Please check the code and try again"
-        case GetCodeStatus.CodeAlreadyUsed => "Your code has already been redeemed"
-      }
-    } yield ()
+    RedemptionCode(inputCode)
+      .leftMap(_ => CodeMalformed)
+      .map {
+        redemptionCode =>
+          for {
+            giftValidationResult <- giftValidator.validate(redemptionCode, "")
+            mergedResult <- if (giftValidationResult == CodeNotFound)
```

Comment:
I'm always slightly wary of using the "try one thing and if it fails try another" logic because it takes away the certainty about what used to happen, if it were to start failing, or anything about which code path it follows if something goes wrong.

Also it perhaps adds a special case where CodeNotFound is not an InvalidCode.  I was trying to think of an alternative name for InvalidCode to make that clearer, then I was unsure why it was like that.

The alternative would be to inspect the first character of the code, so the codes stop being opaque but we do gain the determinism.  (the codes are not actually opaque anyway, as the RedemptionCode type asserts the overall format).  What you you think?

Reply:
I really don't want to rely on parsing strings to work out what type of code we have. If we think that it is too much of a risk to try to validate with two separate systems then I would rather have a separate url for each redemption type. As you've noted elsewhere `CodeNotFound` is actually not an `InvalidCode` for this reason.

Reply:
let's leave it as the try one then the other, I accept your reasons for not wanting to look inside the code and I think that having a separate url would be similar complexity on the backend, only more confusing for readers, so overall worse.